src/main/java/frc/robot/subsystems/vision/SimulatedVision.java

```diff
@@ -0,0 +1,178 @@
+package frc.robot.subsystems.vision;
+
+import edu.wpi.first.math.geometry.*;
+import edu.wpi.first.networktables.NetworkTable;
+import frc.robot.subsystems.vision.VisionConstants.Limelight;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Supplier;
+import org.littletonrobotics.junction.Logger;
+import org.photonvision.simulation.PhotonCameraSim;
+import org.photonvision.simulation.SimCameraProperties;
+import org.photonvision.simulation.VisionSystemSim;
+import org.photonvision.targeting.PhotonPipelineResult;
+
+// Simulate the vision system.
+// Please see the following link for example code
+// https://github.com/PhotonVision/photonvision/blob/2a6fa1b6ac81f239c59d724da5339f608897c510/photonlib-java-examples/swervedriveposeestsim/src/main/java/frc/robot/Vision.java
+public class SimulatedVision extends PhysicalVision {
+  // private final PhotonCamera shooterCamera = new PhotonCamera("shooterCamera");
+  // private final PhotonCamera frontLeftCamera = new PhotonCamera("frontLeftCamera");
+  // private final PhotonCamera frontRightCamera = new PhotonCamera("frontRightCamera");
+  PhotonCameraSim shooterCameraSim;
+  PhotonCameraSim frontLeftCameraSim;
+  PhotonCameraSim frontRightCameraSim;
+  private final VisionSystemSim visionSim;
+  private final Supplier<Pose2d> robotSimulationPose;
+
+  private final int kResWidth = 1280;
+  private final int kResHeight = 800;
+
+  private int[] tagCount = new int[Limelight.values().length];
+
+  public SimulatedVision(Supplier<Pose2d> robotActualPoseInSimulationSupplier) {
+    super();
+    this.robotSimulationPose = robotActualPoseInSimulationSupplier;
+    // Create the vision system simulation which handles cameras and targets on the
+    // field.
+    visionSim = new VisionSystemSim("main");
+
+    // Add all the AprilTags inside the tag layout as visible targets to this
+    // simulated field.
+    visionSim.addAprilTags(VisionConstants.FIELD_LAYOUT);
+    // visionSim.addVisionTargets(TargetModel.kAprilTag36h11);
+
+    // Create simulated camera properties. These can be set to mimic your actual
+    // camera.
+    var cameraProperties = new SimCameraProperties();
+    // cameraProperties.setCalibration(kResWidth, kResHeight, Rotation2d.fromDegrees(97.7));
+    // cameraProperties.setCalibError(0.35, 0.10);
+    cameraProperties.setFPS(15);
+    cameraProperties.setAvgLatencyMs(20);
+    cameraProperties.setLatencyStdDevMs(5);
+
+    // Create a PhotonCameraSim which will update the linked PhotonCamera's values
+    // with visible
+    // targets.
+    // Instance variables
+    shooterCameraSim =
+        new PhotonCameraSim(
+            Limelight.SHOOTER.getSimulationCamera(Limelight.SHOOTER), cameraProperties);
```

Comment:
I feel like this syntax could be cleaned up. If the method is coming from the shooter limelight, we shouldn't have to pass it in

Reply:
nah im just dumb